server/src/http/instance_data.rs

```diff
@@ -0,0 +1,122 @@
+use crate::metrics::edge_metrics::EdgeInstanceData;
+use prometheus::Registry;
+use tracing::{debug, trace, warn};
+
+#[derive(Debug, Clone)]
+pub struct InstanceDataSender {
+    pub unleash_client: Arc<UnleashClient>,
+    pub token: String,
+    pub base_path: String,
+}
+
+impl InstanceDataSender {
+    pub fn from_args(
+        args: CliArgs,
+        instance_data: Arc<EdgeInstanceData>,
+    ) -> Result<Option<Self>, EdgeError> {
+        match args.mode {
+            EdgeMode::Edge(edge_args) => {
+                let instance_id = instance_data.identifier.clone();
+                Ok(edge_args.tokens.first().map(|token| {
+                    let client_meta_information = ClientMetaInformation {
+                        app_name: args.app_name,
+                        instance_id,
+                    };
+                    let http_client = new_reqwest_client(
+                        edge_args.skip_ssl_verification,
+                        edge_args.client_identity.clone(),
+                        edge_args.upstream_certificate_file.clone(),
+                        Duration::seconds(edge_args.upstream_request_timeout),
+                        Duration::seconds(edge_args.upstream_socket_timeout),
+                        client_meta_information.clone(),
+                    )
+                    .expect("Could not construct reqwest client for posting observability data");
+                    let unleash_client = Url::parse(&edge_args.upstream_url.clone())
+                        .map(|url| {
+                            UnleashClient::from_url(
+                                url,
+                                args.token_header.token_header.clone(),
+                                http_client,
+                            )
+                        })
+                        .map(|c| {
+                            c.with_custom_client_headers(edge_args.custom_client_headers.clone())
+                        })
+                        .map(Arc::new)
+                        .map_err(|_| EdgeError::InvalidServerUrl(edge_args.upstream_url.clone()))
+                        .expect("Could not construct UnleashClient");
+                    Self {
+                        unleash_client,
+                        token: token.clone(),
+                        base_path: args.http.base_path.clone(),
+                    }
+                }))
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+pub async fn send_instance_data(
+    instance_data_sender: Option<InstanceDataSender>,
+    prometheus_registry: Registry,
+    our_instance_data: Arc<EdgeInstanceData>,
+    downstream_instance_data: Arc<RwLock<Vec<EdgeInstanceData>>>,
+) {
+    let mut do_the_work = true;
+    loop {
+        let mut empty = true;
+        tokio::time::sleep(std::time::Duration::from_secs(60)).await;
+        if let Some(instance_data_sender) = instance_data_sender.clone() {
```

Comment:
So this function is pretty nested. Again personal opinion, but if we never stored the data in if we couldn't send it, then we could use 

```

        let Some(instance_data_sender) = instance_data_sender.clone() else {
            continue;
        };
```

to flatten this out a little